### ISSUE TITLE
[wish] use local `README.md`when off-line

### ISSUE THREAD

**Issue by retokromer** [amiaopensource/vrecord#249]
I suggest to modify the call of the [documentation](https://github.com/amiaopensource/vrecord/blob/ebc8b142fa30f7ddb0624326a0625b475e72ea80/vrecord#L194) like we did e.g. for [ffmprovisr](https://github.com/amiaopensource/ffmprovisr/blob/58aa0549ff607f3cfbcb0d92404b5959598e9c5c/scripts/ffmprovisr#L11):
- if on-line, then show the version on GitHub
- if off-line, then use the version installed on the computer

**Comment by aeschweik** [comment]
Hi @retokromer, are you planning to work on this?

**Comment by retokromer** [comment]
@aeschweik No, at least not in the near future. In my opinion, we should wait until the documentation has been restructured and after that implement the off-line possibility.

**Comment by retokromer** [comment]
Suggested unsuccessfully in https://github.com/amiaopensource/vrecord/issues/334#issuecomment-397628373, therefore closing.

**Comment by privatezero** [comment]
Hi @retokromer I looked at this a bit today to think of some options to make this the most usable offline - I am leaning towards having a conversion of the docs to html that could be opened in the same manner as the online version? I suppose this could either be done on the github side and then installed, or built more dynamically after install (although that would expand the dependencies to include pandoc or something). Did you have any ideas/preferences for this?

Also, as a related issue, in testing some conversions I noted that the docs available at https://github.com/amiaopensource/vrecord/blob/master/Resources/Documentation/analog_digitization.md include unsourced external pictures. I think we might want to replace these with local images, or at least provide attribution at a minimum.

**Comment by retokromer** [comment]
Thank you very much indeed, @privatezero!

> I am leaning towards having a conversion of the docs to html that could be opened in the same manner as the online version? I suppose this could either be done on the github side and then installed, or built more dynamically after install (although that would expand the dependencies to include pandoc or something). Did you have any ideas/preferences for this?

No, I do not have any preferences.

Regarding ideas, I guess, I would try do generate the pages during the installation and not to have more versions on GitHub, because only one version is easier to maintain. My first try would indeed be pandoc. Yet, of course, if you know «something» more efficient, then I will support it. I was told that Lua is a good language for this kind of work.

>I think we might want to replace these with local images

In my opinion, this would be the easiest solution to implement – and to maintain.

**Comment by retokromer** [comment]
Question: Is anybody working on this?

Context: On my end, I am still convinced that having the documentation stored locally on the computer is useful. And not only for people having unreliable web connection, but also for people using less powerful computers, as they can switch off the network and save resources.

**Comment by privatezero** [comment]
I agree that would be useful! I suppose the easiest way about it would be to either add pandoc as a dependency and have vrecord convert its docs for a local version, or to modify the install formula with either wget or curl to download a local copy of the docs on install? (I don't recall if either of those are installed from the beginning on macOS - so this still might add a dependency)

Thoughts? I'd be happy to poke at this later!

**Comment by retokromer** [comment]
(`pandoc` is not installed on macOS by default, but a dependency in the formula works fine.)

**Comment by retokromer** [comment]
(I have a little preference for `curl` over `wget`, but that‘s probably only my personal taste.)

**Comment by privatezero** [comment]
@retokromer I have started to make some changes to the install formula to get a localized version of the docs at https://github.com/amiaopensource/homebrew-amiaos/blob/changes-for-doc-install/vrecord.rb

It concats all the various doc files into one file, attempts to rebuild the index and then pandocs it into an html doc. So far the index works for sub-headings, but not for the top level headings.

The docs being split into separate files works well online, but I found it counterintuitive for local use - thus the concatenation. After the install formula is changed, work can be done to edit vrecord itself to access these docs. What do you think?

**Comment by retokromer** [comment]
>I have started to make some changes to the install formula to get a localized version of the docs at https://github.com/amiaopensource/homebrew-amiaos/blob/changes-for-doc-install/vrecord.rb

A quick first answer: I guess, there was a copy followed by a double paste… Only the first occurrence (with the full alphabetic list of de dependencies) should be kept.

**Comment by privatezero** [comment]
ha! That would be me pasting out of my test script into a branch! I'll clean that up.

**Comment by privatezero** [comment]
@retokromer, any more thoughts on this approach?

**Comment by retokromer** [comment]
@privatezero I am very fine with this approach! I will try to fire up a macOs and a Linux machine this “Apollo” weekend, and test the current state.

**Comment by retokromer** [comment]
@privatezero I installed via https://github.com/amiaopensource/homebrew-amiaos/blob/changes-for-doc-install/vrecord.rb on macOS (10.14.5) and when opening the GUI I got this warning:

    2019-07-20 10:03:16.317 gtkdialog[68661:752964] *** WARNING: Method userSpaceSca
    leFactor in class NSView is deprecated on 10.7 and later. It should not be used 
    in new applications. Use convertRectToBacking: instead.

Yet much more importantly, when WiFi is disabled `vrecord` tries to open the online version with the browser, what of course fails.

**Comment by privatezero** [comment]
Hi @retokromer, oh yes, I am sure that vrecord would not reflect any of the changes yet as I was only experimenting with changes to the install formula to concat/generate local docs. I figured that once the install formula was worked out satisfactorily, then vrecord could be updated to reflect local paths - sorry for not being clearer!

The changes I was testing revolve around the ugly command in the post_install https://github.com/amiaopensource/homebrew-amiaos/blob/changes-for-doc-install/vrecord.rb#L42, and create locally installed docs within the vrecord directory (must be accessed manually for testing).

**Comment by retokromer** [comment]
@privatezero Thank you for the answer! Sorry, I cannot remember how I tested this, but I will run another test soon.

**Comment by retokromer** [comment]
@privatezero I tested again. The page `Resources/Documentation/vrecord-docs.html` is generated, but the GUI tries always to open the online version with the browser (what of course fails when WiFi is disabled), instead of the local version. 

In [ffmprovisr](https://github.com/amiaopensource/ffmprovisr/blob/gh-pages/scripts/ffmprovisr) or in [cable-bible](https://github.com/amiaopensource/cable-bible/blob/master/cable-bible) there is a “switch“ which allows to consult both online and (with a few restrictions) offline.

**Comment by privatezero** [comment]
@retokromer yeah, so far I had the sub-header links working for the most part, but not the top level headers. Definitely imperfect, but since the source material is markdown as well as split into multiple files, to get it into a usable local html has been a little wonky.

Some of the ugly command I was testing that generates the html is an attempt to do some minor parsing to to get anchors to work better, but not 100% of the way there yet.

**Comment by retokromer** [comment]
@privatezero Thanks again! As said, during the weekend I should have the time to delve more into this. I never have used Markdown, but in the past I parsed a lot of other formats, including some lightweight markup language.

**Comment by dericed** [comment]
Going to close. If interesting to someone, feel welcome to re-open or contribute a pull request.

**Comment by retokromer** [comment]
Hmm… Does inclusivity of this project mean to exclude people with poor internet connections form easy access to the documentation? In my personal opinion, @privatezero already went quite far. I don’t have the rights to re-open this.

**Comment by bturkus** [comment]
Honestly @retokromer I think it's a little unfair of you to connect the closing of this issue to some sort of larger exclusionary sentiment. Yes, progress was made on this, but the conversation has been largely defunct for six months, and nothing at all is preventing you or someone else from (a) opening a new issue and referencing this one, (b) requesting that one of the moderators of the repo re-open the issue, or (c) doing the work and sending in a pull request.

**Comment by retokromer** [comment]
[As said:](https://github.com/amiaopensource/vrecord/issues/249#issuecomment-502438830) On my end, I am still convinced that having the documentation stored locally on the computer is useful. And not only for people having unreliable web connection, but also for people using less powerful computers, as they can switch off the network and save resources.

@bturkus I cannot understand why you consider my point of view unfair. For example in Vientiane, Kathmandu or in Kabul it was actually almost impossible to consult the very useful `vrecord` documentation.

**Comment by dericed** [comment]
> Does inclusivity of this project mean to exclude people with poor internet connections form easy access to the documentation?

I'm uncertain how to answer this as written. Regardless of the project or the community, inclusivity is an intention to reduce exclusion, so I suppose the answer is no.

> @bturkus I cannot understand why you consider my point of view unfair.

I think there is consensus here that access to documentation is useful, but [your earlier comment](https://github.com/amiaopensource/vrecord/issues/249#issuecomment-584508763) feels to me to associate closing a stale issue with an intent to act in an exclusionary way, which I also find a little unfair. I think @bturkus's example reactions of a, b, and c work well for any contributor or onlooker to move this issue forward.

Given the discussion I'll reopen the issue for now, but am considering adding a bot to the repository to flag, nudge, and conditionally close long stale issues.

**Comment by retokromer** [comment]
>I'll reopen the issue for now

Thank you!

**Comment by privatezero** [comment]
For progress to be made on this I think that most work will have to be done on the Homebrew formula to locally generate files, rather than within Vrecord. I made it about as far as I think I can in https://github.com/amiaopensource/homebrew-amiaos/tree/changes-for-doc-install (Not perfect, but mostly functional - currently out of date though).

Since the doc files are split up and decentralized for vrecord, it makes it tricky to have a well formatted/usable local doc. I honestly don't remember why we split the docs into separate files in the past, but recombining them would also help with this issue.

**Comment by dericed** [comment]
Thanks @privatezero, if that's the case, Reto do you mind starting an issue in https://github.com/amiaopensource/homebrew-amiaos and referencing it here.
